examples/barotropic_tide.jl

```diff
@@ -0,0 +1,285 @@
+# of ``z`` at the bottom at the cell centers.
+#
+# In this example we'd like to have a small Gaussian hill at the center of the domain.
+
+# ```math
+# h(x) = -H + h_0 \exp(-x^2 / 2σ^2)
+# ```
+
+h₀ = 50 # m
+width = 5kilometers
+hill(x) = h₀ * exp(-x^2 / 2width^2)
+bottom(x, y) = - H + hill(x)
+
+grid = ImmersedBoundaryGrid(underlying_grid, GridFittedBottom(bottom))
+
+# The hill is small; here's how it looks (note that we don't plot all the way to the ocean surface).
+
+xC = xnodes(grid, Center())
+bottom_boundary = grid.immersed_boundary.bottom_height[1:Nx, 1]
+
+using CairoMakie
+
+fig = Figure(resolution = (700, 200))
+ax = Axis(fig[1, 1],
+          xlabel="x [km]",
+          ylabel="z [m]",
+          limits=((-Lx/2e3, Lx/2e3), (-H, -4H/5)))
+
+lines!(ax, xC/1e3, bottom_boundary)
+
+fig
+
+# Now we want to add a barotropic tide forcing. For example, to add the ``M_2`` tidal forcing
+# we need to add forcing in the ``u``-momentum equation of the form:
+# ```math
+# \partial_t u = \dotsb + F_0 \sin(\omega_2 t)
+# ```
+# where ``\omega_2 = 2π / T_2``, with ``T_2 = 12.421 \,\mathrm{hours}`` the period of the ``M_2`` tide.
+
+# The excursion parameter is a nondimensional number that expresses the ratio of the flow movement
+# due to the tide compared to the size of the width of the hill.
+#
+# ```math
+# \epsilon = \frac{U_{\mathrm{tidal}} / \omega_2}{\sigma}
+# ```
+# 
+# We prescribe the excursion parameter which, in turn, implies a tidal velocity ``U_{\mathrm{tidal}}``
+# which then allows us to determing the tidal forcing amplitude ``F_0``. For the last step, we
+# use Fourier decomposition on the inviscid, linearized momentum equations to determine the
+# flow response for a given tidal forcing. Doing so we get that for the sinusoidal forcing above,
+# the tidal velocity and tidal forcing amplitudes are related via:
+#
+# ```math
+# U_{\mathrm{tidal}} = \frac{\omega_2}{\omega_2^2 - f^2}F_0
+# ```
+#
+# Now we have the way to find the value of the tidal forcing amplitude that would correspond to a
+# given excursion parameter. The Coriolis frequency is needed, so we start by constructing a Coriolis on an ``f``-plane at the
+# mid-latitudes.
+
+coriolis = FPlane(latitude = -45)
+
+# Now we have everything we require to construct the tidal forcing given a value of the
+# excursion parameter.
+#
+T₂ = 12.421hours
+const ω₂ = 2π / T₂ # radians/sec
+
+ε = 0.25 # the excursion parameter
+
+U_tidal = ε * ω₂ * width
+
+const tidal_forcing_amplitude = U_tidal * (ω₂^2 - coriolis.f^2) / ω₂
+
+@inline tidal_forcing(x, y, z, t) = tidal_forcing_amplitude * sin(ω₂ * t)
+nothing #hide
+
+# ## Model
+
+# We built a `HydrostaticFreeSurfaceModel` with an `ImplicitFreeSurface` solver.
+
+using Oceananigans.Models.HydrostaticFreeSurfaceModels: FFTImplicitFreeSurfaceSolver
+
+fft_preconditioner = FFTImplicitFreeSurfaceSolver(grid)
+free_surface = ImplicitFreeSurface(solver_method=:PreconditionedConjugateGradient, preconditioner=fft_preconditioner)
```

Comment:
`SplitExplicitFreeSurface(; grid, cfl = 0.7, max_Δt = 3minutes)` would give you about 6 barotropic steps, so probably it's faster

Reply:
There are a few bugs that didn't allow me to put SplitExplicit here the way you suggest. E.g, substep is not ensured to be integer.

#3131 deals with it; could you have a look there @simone-silvestri?

Reply:
@simone-silvestri 

```julia
julia> free_surface = SplitExplicitFreeSurface(; grid, cfl = 0.7, max_Δt)
SplitExplicitFreeSurface with 51895 steps
```

Reply:
this happens, I believe, because for a Flat dimension, e.g., like y here

```Julia
julia> minimum_yspacing(grid)
1.0
```

Reply:
cc @glwagner remember why we made this choice? I believe there was a reason...

we should either change `minimum_ψspacing` for flat ψ coordinates or ensure the the `substeps` computation in the SplitExplicit takes this into account

Reply:
have a look at b395445

Reply:
omg, with the fix the SplitExplicit rips!

Reply:
@djlikesdjs we should change your experiments to use this free surface solver -- I can help you do it

Reply:
It's unfortunate that the split explicit free surface always has to be configured (ie its not possible to make it default...)